src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncValueTaskMethodBuilder.cs

```diff
@@ -142,44 +127,43 @@ public struct AsyncValueTaskMethodBuilder<TResult>
 
         /// <summary>Associates the builder with the specified state machine.</summary>
         /// <param name="stateMachine">The state machine instance to associate with the builder.</param>
-        public void SetStateMachine(IAsyncStateMachine stateMachine) => _methodBuilder.SetStateMachine(stateMachine);
+        public void SetStateMachine(IAsyncStateMachine stateMachine)
+            => AsyncMethodBuilderCore.SetStateMachine(stateMachine, m_task);
 
         /// <summary>Marks the task as successfully completed.</summary>
         /// <param name="result">The result to use to complete the task.</param>
         public void SetResult(TResult result)
         {
-            if (_useBuilder)
+            if (m_task is null)
             {
-                _methodBuilder.SetResult(result);
+                _result = result;
+                m_task = s_haveResultSentinel;
             }
             else
             {
-                _result = result;
-                _haveResult = true;
+                AsyncMethodBuilderCore.SetExistingTaskResult(m_task, result);
             }
         }
 
         /// <summary>Marks the task as failed and binds the specified exception to the task.</summary>
         /// <param name="exception">The exception to bind to the task.</param>
-        public void SetException(Exception exception) => _methodBuilder.SetException(exception);
+        public void SetException(Exception exception)
+            => AsyncMethodBuilderCore.SetException(ref m_task, exception);
 
         /// <summary>Gets the task for this builder.</summary>
         public ValueTask<TResult> Task
         {
             get
             {
-                if (_haveResult)
-                {
-                    return new ValueTask<TResult>(_result);
-                }
-                else
-                {
-                    _useBuilder = true;
-                    return new ValueTask<TResult>(_methodBuilder.Task);
-                }
+                return ReferenceEquals(s_haveResultSentinel, m_task) ?
+                    new ValueTask<TResult>(_result) :
+                    CreateValueTask(ref m_task);
             }
         }
 
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static ValueTask<TResult> CreateValueTask(ref Task<TResult> task) => new ValueTask<TResult>(task ?? AsyncMethodBuilderCore.InitializeTaskAsPromise(ref task!)); // TODO-NULLABLE: Remove ! when nullable attributes are respected
```

Comment:
Same comment as Same as code in https://github.com/dotnet/coreclr/pull/24431/files#diff-b2f7a19e7ee56b16cfe87d2b115e0059R114